nncf/quantization/algorithms/weight_compression/torch_fx_backend.py

```diff
@@ -0,0 +1,273 @@
+
+    def set_weight(
+        self,
+        node_with_weight: NNCFNode,
+        weight_port_id: int,
+        model: torch.fx.GraphModule,
+        graph: NNCFGraph,
+        weight: Tensor,
+    ) -> None:
+        constant_update_transformation_builder(node_with_weight, weight.data, input_port_id=weight_port_id)(model)
+
+    def insert_adapters(
+        self, wc_params: WeightCompressionParameters, lora_A: Tensor, lora_B: Tensor, int8_lora: bool
+    ) -> None:
+        pass
+
+    def transform_model(
+        self,
+        model: torch.fx.GraphModule,
+        graph: NNCFGraph,
+        weight_compression_parameters: Iterable[WeightCompressionParameters],
+        precomputed_scales: Dict[str, Tensor] = None,
+        precomputed_zero_points: Dict[str, Tensor] = None,
+        lora_correction_algo: LoraCorrectionAlgorithm = None,
+    ) -> torch.fx.GraphModule:
+        transformation_layout = TransformationLayout()
+
+        for wc_params in weight_compression_parameters:
+            compression_config = wc_params.compression_config
+            if compression_config.mode not in [
+                CompressWeightsMode.INT8_ASYM,
+                CompressWeightsMode.INT8_SYM,
+                CompressWeightsMode.INT8,
+            ]:
+                raise ValueError(f"{compression_config.mode.value} is not supported.")
+            weight_node = get_const_node(wc_params.node_with_weight, wc_params.weight_port_id, graph)
+            weight_name = weight_node.node_name
+            consumer_nodes = graph.get_next_nodes(weight_node)
+            weight = self.get_weight(wc_params.node_with_weight, wc_params.weight_port_id, model, graph)
+            if weight is None or not isinstance(weight, Tensor):
+                raise nncf.InternalError(f"Could not find a nncf.tensor in the model by name {weight_name}.")
+
+            # calculates compressed weights and decompression parameters
+            compressed_weight = compress_weight(
+                weight,
+                wc_params.reduction_axes,
+                compression_config,
+                None if precomputed_scales is None else precomputed_scales.get(wc_params.weight_name),
+                None if precomputed_zero_points is None else precomputed_zero_points.get(wc_params.weight_name),
+            )
+            compressed_weight.scale = compressed_weight.scale.astype(dtype=TensorDataType.float16)
+
+            # pack compressed tensor
+            if compression_config.mode == CompressWeightsMode.INT8_SYM:
+                dtype = TensorDataType.int8
+            else:
+                dtype = TensorDataType.uint8
+            packed_tensor = compressed_weight.tensor.astype(dtype)
+
+            self.set_weight(wc_params.node_with_weight, wc_params.weight_port_id, model, graph, packed_tensor)
+
+            if len(consumer_nodes) > 1:
+                raise nncf.InternalError("Shared weights not supported in compression for TorchFX models")
```

Comment:
What is the reason for the restriction?

Reply:
So far using `capture_pre_autograd_graph()` adds an extra node for the shared weights which point to the original weight under the hood. But new nodes are created for them in the graph. I had added this for a scenario where an edge case is ever hit or we change our method of creating torch fx graph where this can be extended.

Reply:
Done, Added support for shared weights in the update constant transformation.

Reply:
Yes, the test cases `test_compress_weights_shared` and `test_compress_weights_model_size_conv` in `tests/torch/fx/test_compress_weights.py` cover this case for compression of shared weights.